nss/integration-tests/nss_test.go

```diff
@@ -0,0 +1,176 @@
+		"Error when getting shadow by name and entry does not exist": {db: "shadow", key: "doesnotexit", wantErr: true},
+
+		"Error when getting passwd by id and entry does not exist": {db: "passwd", key: "404", wantErr: true},
+		"Error when getting group by id and entry does not exist":  {db: "group", key: "404", wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			if tc.cacheDB == "" {
+				tc.cacheDB = "multiple_users_and_groups"
+			}
+
+			var socketPath string
+			if !tc.noDaemon && !tc.noCustomSocket {
+				ctx, cancel := context.WithCancel(context.Background())
+				socketPath = runDaemon(ctx, t, tc.cacheDB)
+				t.Cleanup(cancel)
+			}
+
+			cmds := []string{"getent", tc.db}
+			if tc.key != "" {
+				cmds = append(cmds, tc.key)
+			}
+
+			got, err := outNSSCommandForLib(t, socketPath, originOuts[tc.db], cmds...)
+
+			// Some tests use corrupted databases to trigger specific errors. When opening a corrupted database,
+			// the cache will schedule a clear-up routine to fix the database, so we need to give it some time to
+			// finish before proceeding with the test, otherwise the test cleanup functions would run (removing
+			// the files and directories) before the routine has finished and it will panic.
+			if tc.cacheDB != "multiple_users_and_groups" {
+				time.Sleep(time.Second)
+			}
+
+			if tc.wantErr {
+				require.Error(t, err, "Expected an error, but got none")
+				return
+			}
+			require.NoError(t, err, "Expected no error, but got %v", err)
+
+			want := testutils.LoadWithUpdateFromGolden(t, got)
+			require.Equal(t, want, got, "Outputs must match")
+		})
+	}
+}
+
+func TestMain(m *testing.M) {
+	testutils.InstallUpdateFlag()
+	flag.Parse()
+
+	execPath, cleanup, err := buildDaemon()
+	if err != nil {
+		log.Printf("Setup: failed to build daemon: %v", err)
+		os.Exit(1)
+	}
+	defer cleanup()
+	daemonPath = execPath
+
+	// Creates the directory to store the coverage files for the integration tests.
+	if testutils.WantCoverage() {
+		rawCovDir = os.Getenv("GOCOVERDIR")
+		if rawCovDir == "" {
+			dir, err := os.MkdirTemp("", "authd-coverage")
+			if err != nil {
+				log.Printf("Setup: failed to create temp dir for coverage: %v", err)
+				cleanup()
+				os.Exit(24)
+			}
+			defer os.RemoveAll(dir)
+			rawCovDir = dir
+		}
+	}
+
+	code := m.Run()
+	if code == 0 && rawCovDir != "" {
+		coverprofile := filepath.Join(rawCovDir, "integration-tests.coverprofile")
+		// #nosec:G204 - we control the command arguments in tests
+		cmd := exec.Command("go", "tool", "covdata", "textfmt", fmt.Sprintf("-i=%s", rawCovDir), fmt.Sprintf("-o=%s", coverprofile))
+		if err := cmd.Run(); err != nil {
+			log.Printf("Teardown: failed to parse coverage files: %v", err)
+			cleanup()
+			os.RemoveAll(rawCovDir)
+			os.Exit(24)
+		}
+		testutils.MarkCoverageForMerging(coverprofile)
```

Comment:
Oh, I was thinking the other way around, let’s have all coverage reported in the Go native format (or a form compatible with it), and let it handle the merge. Is it doable? (https://go.dev/blog/integration-test-coverage). That way, we don’t have to track and handle all of this ourself, which would be a great bonus and dropping a lot of code!

Reply:
AFAIK, it's not possible. We have three coverage outputs being generated by the tests:
1) Text format generated by using `--coverprofile=cov.out` when running the tests;
2) Raw format generated by compiling the daemon with the `-cover` flag;
3) Raw format generated by compiling the NSS module with instrumentation;

In order to turn the raw files generated by Go into the textfmt, we need to use the command `go tool covdata` to do the conversion (similar to what `grcov` does with the raw files in Rust).

Reply:
As said, we will have another look, but let’s get that rusting too much in that repo